### PR TITLE
Dns resolve fuzz bug

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
Commit Message: The DNS resolver takes reference to a local variable as its member dns_resolver_options_, and may access invalid data during a DNS refresh, fix by copy the parameter in ctor of DnsResolverImpl.(https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=35272)
Additional Description:NA
Risk Level:low
Testing:add fuzz corpus
Docs Changes:NA
Release Notes:NA
Platform Specific Features:NA
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by asraa** [comment]
Could you please add the link to the bug in Fixes description?

**Review by asraa — COMMENTED** [review]
Is there an easy way to test the refresh?

**Comment by chaoqin-li1123** [comment]
I don't know how to reproduce dns refresh in an easier way, we do have a few related fuzz bugs(35136, 34907) caused by this issue tho.

**Review by asraa — APPROVED** [review]

**Review by asraa — COMMENTED** [review]
@ggreenway can you please take a final pass? (or if you have any ideas for testing/or if this makes sense to you)

**Comment by chaoqin-li1123** [comment]
> I **think** that an integration test with a dns cluster, pointed at a non-existent tcp resolver, will reproduce the issue in an ASAN build. It looks like currently the options are usually on the stack of the caller. They are used the first time in `DnsResolverImpl::DnsResolverImpl` while the stack-based config is still in scope, but then they can be used again `if (status == ARES_ECONNREFUSED)`. Can you try to add an integration test that fails without your fix?
> 
> /wait

we can make the dns_resolver_options in DnsImplTest a variable on stack to create a regression asan test.

File: test/common/network/dns_impl_test.cc [hunk 1]
```diff
@@ -441,11 +441,12 @@ class DnsImplTest : public testing::TestWithParam<Address::IpVersion> {
         Network::Test::getCanonicalLoopbackAddress(GetParam()));
     listener_ = dispatcher_->createListener(socket_, *server_, true);
     updateDnsResolverOptions();
+    envoy::config::core::v3::DnsResolverOptions dns_resolver_options = dns_resolver_options_;
```
Comment (ggreenway) [inline test/common/network/dns_impl_test.cc:444]: Can you add a comment here explaining what we're testing by putting this on the stack?

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait

File: test/common/network/dns_impl_test.cc (hunk 1 above)
Comment (chaoqin-li1123) [reply test/common/network/dns_impl_test.cc:445]: makes sense.

**Review by ggreenway — APPROVED** [review]
Thanks!

**Pull request merged**